client-spark/common/src/main/java/org/apache/spark/shuffle/writer/WriteBufferManager.java

```diff
@@ -170,35 +174,59 @@ public List<ShuffleBlockInfo> addRecord(int partitionId, Object key, Object valu
     if (serializedDataLength == 0) {
       return null;
     }
-    List<ShuffleBlockInfo> result = Lists.newArrayList();
+
+    List<ShuffleBlockInfo> sentBlocks = Lists.newArrayList();
+    requestMemoryAndInsert(partitionId, serializedData, serializedDataLength, sentBlocks);
+    shuffleWriteMetrics.incRecordsWritten(1L);
+
+    if (usedBytes.get() - inSendListBytes.get() > spillSize) {
+      sentBlocks.addAll(clear());
+    }
+    writeTime += System.currentTimeMillis() - start;
+    return sentBlocks;
+  }
+
+  /**
+   * Before inserting a record into its corresponding buffer, the system should check if there is
+   * sufficient buffer memory available. If there isn't enough memory, it will request additional
+   * memory from the {@link TaskMemoryManager}. In the event that the JVM is low on memory,
+   * a spill operation will be triggered. If any memory consumer managed by the {@link TaskMemoryManager}
+   * fails to meet its memory requirements, it will also be triggered one by one.
+   *
+   * If the current buffer manager requests memory and triggers a spill operation,
+   * the buffer that is currently being held should be dropped, and then re-inserted.
+   */
+  private void requestMemoryAndInsert(int partitionId, byte[] serializedData, int serializedDataLength,
+      List<ShuffleBlockInfo> sentBlocks) {
+    long required = Math.max(bufferSegmentSize, serializedDataLength);
     if (buffers.containsKey(partitionId)) {
       WriterBuffer wb = buffers.get(partitionId);
       if (wb.askForMemory(serializedDataLength)) {
-        requestMemory(Math.max(bufferSegmentSize, serializedDataLength));
+        requestMemory(required);
+        if (wb.isAvailable()) {
+          // The readonly of writer buffer means the data has been sent when spilling. To avoid data lost,
```

Comment:
The comment isn't clear enough. You should explain more information.

Reply:
Got it.